src/util-ja3.c

```diff
@@ -0,0 +1,290 @@
+
+    return buffer;
+}
+
+/**
+ * \brief Free allocated buffer.
+ *
+ * \param buffer The buffer to free.
+ */
+void Ja3BufferFree(JA3Buffer *buffer)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be NULL");
+        return;
+    }
+
+    if (buffer->data != NULL) {
+        SCFree(buffer->data);
+    }
+
+    SCFree(buffer);
+}
+
+/**
+ * \internal
+ * \brief Resize buffer if it is full.
+ *
+ * \param buffer The buffer.
+ * \param len    The length of the data that should fit into the buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+static int Ja3BufferResizeIfFull(JA3Buffer *buffer, uint32_t len)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be empty");
+        return -1;
+    }
+
+    if (len == 0) {
+        return 0;
+    }
+
+    while (buffer->used + len + 2 > buffer->size)
+    {
+        buffer->size *= 2;
+        char *tmp = SCRealloc(buffer->data, buffer->size * sizeof(char));
+        if (tmp == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "Error resizing JA3 buffer");
+            return -1;
+        }
+        buffer->data = tmp;
+    }
+
+    return 0;
+}
+
+/**
+ * \brief Append buffer to buffer.
+ *
+ * Append the second buffer to the first and then free it.
+ *
+ * \param buffer1 The first buffer.
+ * \param buffer2 The second buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+int Ja3BufferAppendBuffer(JA3Buffer *buffer1, JA3Buffer *buffer2)
+{
+    if (buffer1 == NULL || buffer2 == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffers should not be NULL");
+        return -1;
+    }
+
+    /* If buffer1 contains no data, then we just copy the second buffer
+       instead of appending its data. */
+    if (buffer1->data == NULL) {
+        Ja3BufferFree(buffer1);
+        *buffer1 = *buffer2;
+        return 0;
+    }
+
+    int rc = Ja3BufferResizeIfFull(buffer1, buffer2->used);
+    if (rc != 0) {
+        Ja3BufferFree(buffer1);
+        Ja3BufferFree(buffer2);
+        return -1;
+    }
+
+    if (buffer2->used == 0) {
+        buffer1->used += snprintf(buffer1->data + buffer1->used, buffer1->size -
+                                  buffer1->used, ",");
+    } else {
+        buffer1->used += snprintf(buffer1->data + buffer1->used, buffer1->size -
+                                  buffer1->used, ",%s", buffer2->data);
+    }
+
+    Ja3BufferFree(buffer2);
+
+    return 0;
+}
+
+/**
+ * \internal
+ * \brief Return number of digits in number.
+ *
+ * \param num The number.
+ *
+ * \return digits Number of digits.
+ */
+static unsigned NumberOfDigits(unsigned num)
+{
+    if (num < 10) {
+        return 1;
+    }
+
+    return 1 + NumberOfDigits(num / 10);
+}
+
+/**
+ * \brief Add value to buffer.
+ *
+ * \param buffer The buffer.
+ * \param value  The value.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+int Ja3BufferAddValue(JA3Buffer *buffer, int value)
```

Comment:
this function takes the buffer in, then on various error conditions frees it. This leaves it to the called to make sure there is no dangling ptr. I'm not seeing a clean handling of this in the callers, other than returning error codes there.

Reply:
Would it be more clear what's happening if the 'Ja3Buffer' is freed in the callers instead of in 'Ja3BufferAddValue'? This should really only happen if we are out of memory (I think).